presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PropertyDerivations.java

```diff
@@ -183,13 +183,23 @@ public ActualProperties visitEnforceSingleRow(EnforceSingleRowNode node, List<Ac
         public ActualProperties visitAssignUniqueId(AssignUniqueId node, List<ActualProperties> inputProperties)
         {
             ActualProperties properties = Iterables.getOnlyElement(inputProperties);
+
+            ImmutableList.Builder<LocalProperty<Symbol>> localPropertiesBuilder = ImmutableList.builder();
+            localPropertiesBuilder.addAll(properties.getLocalProperties());
+            localPropertiesBuilder.add(new GroupingProperty<>(ImmutableList.of(node.getIdColumn())));
+            node.getSource().getOutputSymbols().stream()
+                    .forEach(column -> localPropertiesBuilder.add(new ConstantProperty<>(column)));
+
             if (properties.getNodePartitioning().isPresent()) {
                 // preserve input (possibly preferred) partitioning
-                return properties;
+                return ActualProperties.builderFrom(properties)
+                        .local(localPropertiesBuilder.build())
+                        .build();
             }
 
             return ActualProperties.builderFrom(properties)
```

Comment:
Using `ActualProperties.builderFrom(properties)` preserves `ActualProperties#constants`, but then we actually build a different hierarchy of local properties so I'm not sure how consistent produced properties are.

Also `ConstantProperty` are always first in local properties hierarchy, so I'm not sure if `GroupingProperty` will be matched when `ConstantProperty` are first on local properties list. Maybe `ContantProperty` should simplify `GroupingProperty`? Then we could just write:
```
return ActualProperties.builder(properties)
  .global(partitionedOn(ARBITRARY_DISTRIBUTION, ImmutableList.of(node.getIdColumn()), Optional.empty()))
  .local(localPropertiesBuilder.build(new GroupingProperty<>(ImmutableList.of(node.getOutputSymbols()))))
  .constants(properties.getConstants())
  .build();
```

Reply:
@sopel39 

> Using ActualProperties.builderFrom(properties) preserves ActualProperties#constants, but then we actually build a different hierarchy of local properties so I'm not sure how consistent produced properties are.

Preserving `ActualProperties#constants` is correct. If some column has a constant value, then adding another column (`unique`) doesn't change that.

> Also ConstantProperty are always first in local properties hierarchy, so I'm not sure if GroupingProperty will be matched when ConstantProperty are first on local properties list.

`ConstantProperty` may appear at any level in local properties hierarchy. `ActualProperties#constants` do get translated into `ConstantProperty` instances at the top levels of the hierarchy. (However, not every `ConstantProperty` can be translates to an entry in `ActualProperties#constants`.)

> I'm not sure if GroupingProperty will be matched when ConstantProperty are first on local properties list. Maybe ConstantProperty should simplify GroupingProperty too?

They will. `GroupingProperty#isSimplifiedBy` returns `true` for `ConstantProperty`.

    @Override
    public boolean isSimplifiedBy(LocalProperty<E> known)
    {
        return known instanceof ConstantProperty || getColumns().containsAll(known.getColumns());
    }

To confirm the above, I added a test to `TestLocalProperties` (I don't plan to add it PR, but let me know if I should):

```
public class TestLocalProperties
{
    @Test
    public void test()
    {
        List<LocalProperty<String>> desired = builder().grouped("unique", "a", "b", "c").build();

        List<LocalProperty<String>> actual = builder()
                .grouped("a")
                .grouped("unique")
                .constant("a")
                .constant("b")
                .constant("c")
                .build();

        assertMatch(actual, desired, Optional.empty());

        actual = builder()
                .constant("a")
                .grouped("unique")
                .constant("a")
                .constant("b")
                .constant("c")
                .build();
        assertMatch(actual, desired, Optional.empty());

        actual = builder()
                .sorted("a", SortOrder.ASC_NULLS_FIRST)
                .grouped("unique")
                .constant("a")
                .constant("b")
                .constant("c")
                .build();
        assertMatch(actual, desired, Optional.empty());
    }
```

Reply:
> Preserving ActualProperties#constants is correct. If some column has a constant value, then adding another column (unique) doesn't change that.

But previous top-level local properties become some nested properties in local properties hierarchy. Is that OK? It's not longer a leading constant (as `ActualProperties#constants` seem to be)

Still I would prefer it to be explicit, e.g:
```
return ActualProperties.builder()
  .global(partitionedOn(ARBITRARY_DISTRIBUTION, ImmutableList.of(node.getIdColumn()), Optional.empty()))
  .constants(properties.getConstants())|.leadingConstants(properties.getLeadingConstants())
  .local(newLocalProperties.build())
```